_blog/2022-12-23-procedural-macros.mdx

```diff
@@ -0,0 +1,417 @@
+```javascript
+> Object.keys({a: 3, b: 2});
+[ "a", "b" ]
+> Object.keys({b: 3, a: 2});
+[ "b", "a" ]
+```
+
+We also use `Reflect.get` to get a property under a given string key (`item[name]` is equivalent).
+
+> The idea of the example is not to show how to do low-level byte conversion in JavaScript but to show how you can mix in the the introspection logic. As we will see later runtime reflection is very difficult to do in Rust as there are no equivalent `Object.entries`, `Reflect.get` functions or a `typeof` operator in the language.
+
+Now we can serialize the array with `arrayToString(countries)`, we want a way to reverse the process! However deserialization process becomes a bit of a problem, reflection in JS can only be done when we have a existing structure in inspect. As there are no type/shape declarations in plain JavaScript, there is no reference of the shape of the object we want to deserialize our serialized string into.
+
+If we were using JSON we would be okay as the keys are embed into the serialized format, in our example we don't save the keys. Instead we can send a representation array of the key type pairs we want the objects to look like. Using `fields` and with a bit of conversion from our low level formatted string we have the following:
+```javascript
+const stringToFloat32 = (string, offset) => {
+	const u16 = new Uint16Array([
+		string.charCodeAt(offset),
+		string.charCodeAt(offset + 1),
+	]).buffer;
+	return new Float32Array(u16)[0];
+};
+
+function arrayFromString(string, fields) {
+	let i = 0;
+	const entries = stringToFloat32(string, i);
+    i += 2;
+	const array = [];
+	for (let arrayIndex = 0; arrayIndex < entries; arrayIndex++) {
+        const object = {};
+		for (const { name, ty } of fields) {
+			let value;
+			switch (ty) {
+				case "string":
+					const length = stringToFloat32(string, i);
+					i += 2;
+					value = String.fromCharCode(
+						...Array.from({ length }, (_, j) => string.charCodeAt(j + i)),
+					);
+					i += length;
+					break;
+				case "number":
+					value = stringToFloat32(string, i);
+					i += 2;
+					break;
+				case "boolean":
+					value = string.charCodeAt(i) === 0;
+					i++;
+					break;
+			}
+			Reflect.set(object, name, value);
+		}
+        array.push(object);
+	}
+	return array;
+}
+```
+
+Here we use another part of reflection `Reflect.set` which allows us to set a property of an existing object based on a string key (`name`).
+### Problems with reflection in JavaScript
+The `arrayFromString` function assumes that the caller has passed a standard array where every object has the same type. `Reflect.get` will fail at runtime as if the property doesn't not exist `null` will be returned and there is a catch to throw an `Error`, so nothing unsafe. The downside is the both the dynamic property lookup and null check is expensive at runtime. 

```

Comment:
Wait, `arrayFromString` uses `Reflect.set` and not `Reflect.get` right?

Reply:
Ah yes, that paragraph is talking about the `arrayToString` function not `arrayFromString` 🤦‍♂️